spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+  private final int defaultParallelism;
+  private final boolean snapshotIdInheritanceEnabled;
+  private final long targetManifestSizeBytes;
+
+  private final Encoder<ManifestFile> manifestEncoder = Encoders.javaSerialization(ManifestFile.class);
+  private final Encoder<Entry> entryEncoder = Encoders.javaSerialization(Entry.class);
+  private final Encoder<Bin> binEncoder = Encoders.bean(Bin.class);
+
+  private Predicate<ManifestFile> predicate = manifest -> true;
+  private String stagingLocation = null;
+
+  RewriteManifestsAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.specs = table.specs();
+    this.summary = Maps.newHashMap();
+    this.defaultParallelism = Integer.parseInt(
+        spark.conf().get("spark.default.parallelism", "200"));
+    this.snapshotIdInheritanceEnabled = PropertyUtil.propertyAsBoolean(
+        table.properties(),
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED,
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+    this.targetManifestSizeBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      fileIO = new HadoopFileIO(conf::value);
+    } else {
+      fileIO = table.io();
+    }
+  }
+
+  public RewriteManifestsAction rewriteIf(Predicate<ManifestFile> newPredicate) {
+    this.predicate = newPredicate;
+    return this;
+  }
+
+  public RewriteManifestsAction stagingLocation(String newStagingLocation) {
+    this.stagingLocation = newStagingLocation;
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsAction set(String property, String value) {
+    summary.put(property, value);
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsActionResult execute() {
+    Preconditions.checkArgument(stagingLocation != null, "Staging location must be set");
+
+    List<ManifestFile> matchingManifests = findMatchingManifests();
+    if (matchingManifests.isEmpty()) {
+      return null;
+    }
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+
+    int parallelism = Math.min(matchingManifests.size(), defaultParallelism);
+    JavaRDD<ManifestFile> manifestRDD = sparkContext.parallelize(matchingManifests, parallelism);
+    Dataset<ManifestFile> manifestDS = spark.createDataset(manifestRDD.rdd(), manifestEncoder);
+    Dataset<Entry> manifestEntryDS = manifestDS.flatMap(toEntries(io, specs), entryEncoder);
+
+    try {
+      manifestEntryDS.cache();
+
+      long manifestEntrySizeBytes = computeManifestEntrySizeBytes(matchingManifests);
```

Comment:
Can we do this using manifest metadata instead? We have `ManifestFile.length` as well as stats for the number of each type of entry. It seems like we could get a fair estimate using those numbers.

The reason why I point this out is that we never recommend caching a dataset in Spark. In environments that use dynamic allocation, it's a mess. That is probably getting better in 3.0 because you can use the shuffle system to cache, but I would still like to avoid caching a dataset possible.